### PR TITLE
Triggers refactor

### PR DESCRIPTION
**Pull request by matis755**

### CODE REVIEW

File: src/trigger_out.h [hunk 1]
```diff
@@ -3,17 +3,19 @@
...
-namespace ara { namespace com {
+#include "sm_types.h"
+
+namespace ara::com {
 
 /** @brief Fills [SWS_SM_91008] */
     class TriggerOut {
     public:
-        bool IsNotifier() const;
+        sm::SMStateType GetSMState() const;
```
Comment (r4jmond) [inline src/trigger_out.h]: Change to GetNotifier

File: src/trigger_in.h [hunk 2]
```diff
@@ -3,17 +3,22 @@
...
+        sm::SMStateType GetDesiredState() const;
 
-        void SetTrigger(bool trigger);
+        void SetTrigger(sm::SMStateType);
+        void DiscardTrigger();
 
     private:
-        bool trigger;
+        bool isTrigger = false;
+        sm::SMStateType trigger;
```
Comment (r4jmond) [inline src/trigger_in.h]: Please create another type for trigger containing sm::SMStateType and isTrigger. We should only have field 'trigger' according to requirements.

File: tests/test_trigger_inout.cpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+/** @brief Fills [SWS_SM_00020]
+ *         Fills [SWS_SM_00021] */
+TEST_F(smTests, testInternalStateInfluenceAndPropagation) {
+    mySM.triggerInOut.SetSMState(ara::sm::SMStateType::On);
+
+    // SM part
+    EXPECT_TRUE(mySM.triggerInOut.IsTrigger());
+    mySM.triggerInOut.SetTransitionResult(ara::sm::ErrorType::kSuccess, ara::sm::SMStateType::On);
+    mySM.triggerInOut.DiscardTrigger();
```
Comment (r4jmond) [inline tests/test_trigger_inout.cpp]: SM part? do you mean that this should be executed from SM?
If so, move it to SM, we will clean SM later

File: tests/test_trigger_out.cpp [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#include "sm_tests.h"
+
+/** @brief Fills [SWS_SM_00020] */
+TEST_F(smTests, testInternalStatePropagation) {
+    mySM.triggerOut.SetNotifier(ara::sm::SMStateType::On);
+    EXPECT_EQ(mySM.triggerOut.GetSMState(), ara::sm::SMStateType::On);
```
Comment (r4jmond) [inline tests/test_trigger_out.cpp]: This test shall set SM to ON. Line 5 should be moved to SM

File: tests/test_trigger_in.cpp [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#include "sm_tests.h"
+
+/** @brief Fills [SWS_SM_00021] */
+TEST_F(smTests, testInternalStateInfluence) {
+    mySM.triggerIn.SetTrigger(ara::sm::SMStateType::Off);
+    EXPECT_EQ(mySM.triggerIn.GetDesiredState(), ara::sm::SMStateType::Off);
+    EXPECT_TRUE(mySM.triggerIn.IsTrigger());
+
+    mySM.triggerIn.DiscardTrigger();
```
Comment (r4jmond) [inline tests/test_trigger_in.cpp]: Move to SM

File: src/trigger_in_out.h [hunk 6]
```diff
@@ -3,22 +3,26 @@
...
     /** @brief Fills [SWS_SM_91009] */
     class TriggerInOut {
     public:
-        bool IsNotifier() const;
-
-        void SetNotifier(bool newNotifier);
+        sm::TriggerInOutNotifierType GetTransitionResult() const;
+        void SetSMState(sm::SMStateType);
 
+        void SetTransitionResult(sm::ErrorType, sm::SMStateType);
```
Comment (r4jmond) [inline src/trigger_in_out.h]: pls name these parameters

File: src/trigger_in_out.h [hunk 7]
```diff
@@ -3,22 +3,26 @@
...
+
+namespace ara::com {
 
     /** @brief Fills [SWS_SM_91009] */
     class TriggerInOut {
     public:
-        bool IsNotifier() const;
-
-        void SetNotifier(bool newNotifier);
+        sm::TriggerInOutNotifierType GetTransitionResult() const;
```
Comment (r4jmond) [inline src/trigger_in_out.h]: Rename to GetNotifier

File: src/trigger_in_out.h [hunk 8]
```diff
@@ -3,22 +3,26 @@
...
+        void SetSMState(sm::SMStateType);
```
Comment (r4jmond) [inline src/trigger_in_out.h]: Rename to SetTrigger

File: src/trigger_in_out.h [hunk 9]
```diff
@@ -3,22 +3,26 @@
...
-        void SetTrigger(bool newTrigger);
+        void DiscardTrigger();
 
     private:
-        bool notifier;
-        bool trigger;
+        sm::TriggerInOutNotifierType notifier;
+        sm::SMStateType trigger;
+
+        sm::ErrorType transitionResult;
```
Comment (r4jmond) [inline src/trigger_in_out.h]: Delete this field, unused

**Review by r4jmond — CHANGES_REQUESTED** [review]

File: src/trigger_in_out.h (hunk 9 above)
Comment (matis755) [reply src/trigger_in_out.h]: Done

File: src/trigger_in_out.h (hunk 8 above)
Comment (matis755) [reply src/trigger_in_out.h]: Done

File: src/trigger_in_out.h (hunk 7 above)
Comment (matis755) [reply src/trigger_in_out.h]: Done

File: src/trigger_in_out.h (hunk 6 above)
Comment (matis755) [reply src/trigger_in_out.h]: Done

File: tests/test_trigger_in.cpp (hunk 5 above)
Comment (matis755) [reply tests/test_trigger_in.cpp]: Done

File: tests/test_trigger_out.cpp (hunk 4 above)
Comment (matis755) [reply tests/test_trigger_out.cpp]: Done

File: tests/test_trigger_inout.cpp (hunk 3 above)
Comment (matis755) [reply tests/test_trigger_inout.cpp]: Done

File: src/trigger_in.h (hunk 2 above)
Comment (matis755) [reply src/trigger_in.h]: Done

File: src/trigger_out.h (hunk 1 above)
Comment (matis755) [reply src/trigger_out.h]: Done

File: src/sm_types.h [hunk 10]
```diff
@@ -130,5 +130,27 @@ namespace ara::sm {
...
+        /** @brief To inform applications about state transition */
+        sm::ErrorType result = ErrorType::kFailed;
+        /** @brief SM state after processing trigger */
+        SMStateType currentSMState = SMStateType::Off;
+    };
+
+    /**
+    * @brief Fills [SWS_SM_91007]
+    * @details Type used for triggering SM state change */
+    class TriggerInMessageType {
```
Comment (r4jmond) [inline src/sm_types.h]: maybe just TriggerType? Can be used both for TriggerIn and TriggerInOut

File: src/trigger_in_out.h [hunk 11]
```diff
@@ -3,22 +3,25 @@
...
-        void SetTrigger(bool newTrigger);
+        void DiscardTrigger();
 
     private:
-        bool notifier;
-        bool trigger;
+        sm::TriggerInOutNotifierType notifier;
+        sm::SMStateType trigger;
+
+        bool isTrigger = false;
```
Comment (r4jmond) [inline src/trigger_in_out.h]: use TriggerType here

**Review by r4jmond — COMMENTED** [review]
Change TriggerInMessageType to TriggerType and use it in both TriggerIn and TriggerInOut

File: tests/test_trigger_out.cpp [hunk 12]
```diff
@@ -0,0 +1,9 @@
+#include "sm_tests.h"
+
+/** @brief Fills [SWS_SM_00020] */
+TEST_F(smTests, testInternalStatePropagation) {
+    // let changes to be applied
```
Comment (r4jmond) [inline tests/test_trigger_out.cpp:6]: What changes? You didn't call SMSetState(on)

**Review by r4jmond — COMMENTED** [review]
please use TriggerType for both TriggerIn and TriggerInOut

File: src/sm_types.h (hunk 10 above)
Comment (matis755) [reply src/sm_types.h]: Done

File: src/trigger_in_out.h (hunk 11 above)
Comment (matis755) [reply src/trigger_in_out.h]: Done

File: tests/test_trigger_out.cpp (hunk 12 above)
Comment (matis755) [reply tests/test_trigger_out.cpp:6]: It's no logic implemented yet in SM. Now, SM sets TriggerOut notifier without any conditions

File: src/trigger_in_out.h [hunk 13]
```diff
@@ -3,22 +3,23 @@
...
     /** @brief Fills [SWS_SM_91009] */
     class TriggerInOut {
     public:
-        bool IsNotifier() const;
-
-        void SetNotifier(bool newNotifier);
+        sm::TriggerInOutNotifierType GetNotifier() const;
+        void SetTrigger(sm::SMStateType);
 
+        void SetTransitionResult(sm::ErrorType result, sm::SMStateType currentSMState);
```
Comment (r4jmond) [inline src/trigger_in_out.h]: change to SetNotifier

Comment (matis755) [reply src/trigger_in_out.h]: Done

**Review by r4jmond — APPROVED** [review]

**Pull request merged**